static/app/components/smartSearchBar/index.tsx

```diff
@@ -749,14 +769,211 @@ class SmartSearchBar extends React.Component<Props, State> {
+      ? this.state.searchGroups
+      : this.state.searchGroups.filter(
+          item => item.value && item.value.indexOf(preparedQuery) !== -1
+        );
+
+    this.setState({
+      searchTerm: query,
+      searchGroups: filteredSearchGroups,
+    });
+
+    const tag = supportedTags[tagName];
+
+    if (!tag) {
+      return {
+        searchItems: [],
+        recentSearchItems: [],
+        tagName,
+        type: 'invalid-tag',
+      };
+    }
+
+    // Ignore the environment tag if the feature is active and
+    // excludeEnvironment = true
+    if (excludeEnvironment && tagName === 'environment') {
+      return null;
+    }
+
+    const fetchTagValuesFn =
+      tag.key === 'firstRelease'
+        ? this.getReleases
+        : tag.predefined
+        ? this.getPredefinedTagValues
+        : this.getTagValues;
+
+    const [tagValues, recentSearches] = await Promise.all([
+      fetchTagValuesFn(tag, preparedQuery),
+      this.getRecentSearches(),
+    ]);
+
+    return {
+      searchItems: tagValues ?? [],
+      recentSearchItems: recentSearches ?? [],
+      tagName: tag.key,
+      type: 'tag-value',
+    };
+  };
+
+  generateOpAutocompleteGroup(
+    validOps: readonly TermOperator[],
+    tagName: string
+  ): AutocompleteGroup {
+    const operatorMap = generateOperatorEntryMap(tagName);
+    const operatorItems = validOps.map(op => operatorMap[op]);
+    return {
+      searchItems: operatorItems,
+      recentSearchItems: undefined,
+      tagName: '',
+      type: 'tag-operator' as ItemType,
+    };
+  }
+
+  showDefaultSearches = async () => {
+    const {query} = this.state;
+    const [defaultSearchItems, defaultRecentItems] = this.props.defaultSearchItems!;
+
+    if (!defaultSearchItems.length) {
+      // Update searchTerm, otherwise <SearchDropdown> will have wrong state
+      // (e.g. if you delete a query, the last letter will be highlighted if `searchTerm`
+      // does not get updated)
+      this.setState({searchTerm: query});
+
+      const tagKeys = this.getTagKeys('');
+      const recentSearches = await this.getRecentSearches();
+
+      this.updateAutoCompleteState(tagKeys, recentSearches ?? [], '', 'tag-key');
+      return;
+    }
+    // cursor on whitespace show default "help" search terms
+    this.setState({searchTerm: ''});
+
+    this.updateAutoCompleteState(defaultSearchItems, defaultRecentItems, '', 'default');
+    return;
+  };
+
+  getCursorToken = (
+    parsedQuery: ParseResult,
+    cursor: number
+  ): ParseResult[number] | null => {
+    // traverse AST to find first matching filter based on cursor position
+    for (const node of parsedQuery) {
+      if (node.type !== Token.Spaces && isWithinToken(node, cursor)) {
+        // traverse into a logic group to find specific filter
+        if (node.type === Token.LogicGroup) {
+          return this.getCursorToken(node.inner, cursor);
+        }
+        return node;
+      }
+    }
+    return null;
```

Comment:
nit but I would rewrite with some guard statements to make it less 'pyramid-y' 

```tsx
for (const node of parsedQuery) {
  if (node.type === Token.Spaces || !isWithinToken(node, cursor)) {
    continue;
  }

  // traverse into a logic group to find specific filter
  if (node.type === Token.LogicGroup) {
    return this.getCursorToken(node.inner, cursor);
  }

  return node;
  }
}

return null;
```

Also, this is my general personal rule, but I often try and put a blank line before returns, since I think it breaks things a part a bit nicer.

Reply:
oooh gotcha sounds good will change